### ISSUE TITLE
Timeout option for parallel DataLoader

### ISSUE THREAD

**Issue by vadimkantorov** [pytorch/pytorch#2474]
At least with `pin_memory = False`, torch uses Python's `SimpleQueue` to implement producer-consumer pattern to sample example indices and then load/preprocess them on worker threads. A problem with `SimpleQueue` is that they use native OS pipes which have small buffer sizes (64Kb on Linux), lock when this buffer size is exceeded and don't provide a way to fail on lock timeout.

When for some reason worker threads are locked or die (e.g. due to failed unpickling) this leads to a permanent hang in `_put_indices { indices_queue.put(...) }` in `DataLoaderIter` (for me it happens even in constructor during batch pre-fetching).

If we could have a way to specify a timeout for `SimpleQueue.put`, this may eliminate a class of uncertain hangs and at least turn them into exceptions. The `put` calls are in `_put_indices` and in `_worker_loop`.

A particularly nasty deadlock may theoretically happen if all worker threads are blocked because `data_queue` exceeded its buffer size, and the main thread instead of removing the results from `data_queue`, fills up `indices_queue` and gets blocked because `indices_queue` is filled up. Given large enough batches (say with large lists in them transferred by value), this might happen during initial batch pre-fetching / priming.

**Comment by colesbury** [comment]
Can you provide sample code that demonstrates the problem? Timeouts are not very attractive because there's no reasonable default timeout (except not having a timeout).

**Comment by vadimkantorov** [comment]
While crafting such an example (that would immediately block worker threads by overflowing the 64Kb SimpleQueue buffer size and successifely blocking the main thread, since no worker threads would remove also large indices from indices_queue), I hit another problem, with exceptions and a deadlock too (but probably due to died worker threads). 

Do you understand where does `_run_finalizers` come from? Why is `reduce_storage` concerned? No tensors are used here.

```python
import torch.utils.data

impl = type('', (), dict(__len__ = lambda self: 100, __iter__ = lambda self: (range(20000) for batch_ind in range(100)), __getitem__ = lambda self, i: range(1000000)))()
d = torch.utils.data.DataLoader(dataset = impl, sampler = impl, num_workers = 4)
for i, x in enumerate(d):
    print(i)
```

```
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/util.py", line 274, in _run_finalizers
  File "/usr/lib/python2.7/multiprocessing/util.py", line 207, in __call__
  File "/usr/lib/python2.7/shutil.py", line 239, in rmtree
  File "/usr/lib/python2.7/shutil.py", line 237, in rmtree
OSError: [Errno 24] Too many open files: '/tmp/pymp-vR3nxX'
Process Process-1:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
  File ".../torch/utils/data/dataloader.py", line 46, in _worker_loop
  File "/usr/lib/python2.7/multiprocessing/queues.py", line 392, in put
  File ".../torch/multiprocessing/queue.py", line 17, in send
  File "/usr/lib/python2.7/pickle.py", line 224, in dump
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 606, in save_list
  File "/usr/lib/python2.7/pickle.py", line 639, in _batch_appends
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 67, in dispatcher
  File "/usr/lib/python2.7/pickle.py", line 401, in save_reduce
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 66, in dispatcher
  File ".../torch/multiprocessing/reductions.py", line 115, in reduce_storage
  File "/usr/lib/python2.7/multiprocessing/reduction.py", line 145, in reduce_handle
OSError: [Errno 24] Too many open files
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/util.py", line 274, in _run_finalizers
  File "/usr/lib/python2.7/multiprocessing/util.py", line 207, in __call__
  File "/usr/lib/python2.7/shutil.py", line 239, in rmtree
  File "/usr/lib/python2.7/shutil.py", line 237, in rmtree
OSError: [Errno 24] Too many open files: '/tmp/pymp-QhgvsZ'
Process Process-2:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
  File ".../torch/utils/data/dataloader.py", line 46, in _worker_loop
  File "/usr/lib/python2.7/multiprocessing/queues.py", line 392, in put
  File ".../torch/multiprocessing/queue.py", line 17, in send
  File "/usr/lib/python2.7/pickle.py", line 224, in dump
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 606, in save_list
  File "/usr/lib/python2.7/pickle.py", line 639, in _batch_appends
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 67, in dispatcher
  File "/usr/lib/python2.7/pickle.py", line 401, in save_reduce
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 66, in dispatcher
  File ".../torch/multiprocessing/reductions.py", line 115, in reduce_storage
  File "/usr/lib/python2.7/multiprocessing/reduction.py", line 145, in reduce_handle
OSError: [Errno 24] Too many open files
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/util.py", line 274, in _run_finalizers
  File "/usr/lib/python2.7/multiprocessing/util.py", line 207, in __call__
  File "/usr/lib/python2.7/shutil.py", line 239, in rmtree
  File "/usr/lib/python2.7/shutil.py", line 237, in rmtree
OSError: [Errno 24] Too many open files: '/tmp/pymp-duhyvc'
Process Process-4:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
  File ".../torch/utils/data/dataloader.py", line 46, in _worker_loop
  File "/usr/lib/python2.7/multiprocessing/queues.py", line 392, in put
  File ".../torch/multiprocessing/queue.py", line 17, in send
  File "/usr/lib/python2.7/pickle.py", line 224, in dump
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 606, in save_list
  File "/usr/lib/python2.7/pickle.py", line 639, in _batch_appends
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 67, in dispatcher
  File "/usr/lib/python2.7/pickle.py", line 401, in save_reduce
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 66, in dispatcher
  File ".../torch/multiprocessing/reductions.py", line 115, in reduce_storage
  File "/usr/lib/python2.7/multiprocessing/reduction.py", line 145, in reduce_handle
OSError: [Errno 24] Too many open files
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/util.py", line 274, in _run_finalizers
  File "/usr/lib/python2.7/multiprocessing/util.py", line 207, in __call__
  File "/usr/lib/python2.7/shutil.py", line 239, in rmtree
  File "/usr/lib/python2.7/shutil.py", line 237, in rmtree
OSError: [Errno 24] Too many open files: '/tmp/pymp-fYDYx_'
Process Process-3:
Traceback (most recent call last):
  File "/usr/lib/python2.7/multiprocessing/process.py", line 258, in _bootstrap
  File "/usr/lib/python2.7/multiprocessing/process.py", line 114, in run
  File ".../torch/utils/data/dataloader.py", line 46, in _worker_loop
  File "/usr/lib/python2.7/multiprocessing/queues.py", line 392, in put
  File ".../torch/multiprocessing/queue.py", line 17, in send
  File "/usr/lib/python2.7/pickle.py", line 224, in dump
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 606, in save_list
  File "/usr/lib/python2.7/pickle.py", line 639, in _batch_appends
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 67, in dispatcher
  File "/usr/lib/python2.7/pickle.py", line 401, in save_reduce
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
  File "/usr/lib/python2.7/pickle.py", line 286, in save
  File "/usr/lib/python2.7/multiprocessing/forking.py", line 66, in dispatcher
  File ".../torch/multiprocessing/reductions.py", line 115, in reduce_storage
  File "/usr/lib/python2.7/multiprocessing/reduction.py", line 145, in reduce_handle
OSError: [Errno 24] Too many open files

HERE IS A DEADLOCK. PRESSING CTRL+C

Traceback (most recent call last):
  File "bug.py", line 7, in <module>
    for i, x in enumerate(d):
  File ".../torch/utils/data/dataloader.py", line 302, in __iter__
    return DataLoaderIter(self)
  File ".../torch/utils/data/dataloader.py", line 172, in __init__
    self._put_indices()
  File ".../torch/utils/data/dataloader.py", line 214, in _put_indices
    self.index_queue.put((self.send_idx, indices))
  File "/usr/lib/python2.7/multiprocessing/queues.py", line 392, in put
    return send(obj)
  File ".../torch/multiprocessing/queue.py", line 17, in send
    ForkingPickler(buf, pickle.HIGHEST_PROTOCOL).dump(obj)
  File "/usr/lib/python2.7/pickle.py", line 224, in dump
    self.save(obj)
  File "/usr/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python2.7/pickle.py", line 554, in save_tuple
    save(element)
  File "/usr/lib/python2.7/pickle.py", line 286, in save
    f(self, obj) # Call unbound method with explicit self
  File "/usr/lib/python2.7/pickle.py", line 606, in save_list
    self._batch_appends(iter(obj))
  File "/usr/lib/python2.7/pickle.py", line 630, in _batch_appends
    x = items.next()
KeyboardInterrupt
```

**Comment by vadimkantorov** [comment]
Even if default timeout is infinite, I think it's good to provide an option, a user may set a reasonable timeout for their task if they wish to face an exception rather than an infinite hang (due to problems in workers).

**Comment by vadimkantorov** [comment]
In my case I got hit by https://github.com/pytorch/pytorch/issues/1838. In https://github.com/opencv/opencv/issues/5150 @ogrisel mentioned they created https://github.com/tomMoral/loky to support "spawn" on Python 2.7 and to detect failed workers.

**Comment by SsnL** [comment]
Timeout was added in #3474

**Comment by vadimkantorov** [comment]
@SsnL Is DataLoader now able to correctly break if https://github.com/pytorch/pytorch/issues/2474#issuecomment-323203974 happens?

**Comment by SsnL** [comment]
We have switched to using `mp.Queue` (not `SimpleQueue`) now. I tested the above code on master. Interestingly, it makes the worker segfault. It works with `num_workers=0`.

I adapted as following to make it run:
```
import torch.utils.data

class Dataset(object):
  def __len__(self):
    return 100

  def __getitem__(self, i):
    return list(range(100000))


class Sampler(torch.utils.data.Sampler):
  def __iter__(self):
    return (range(100000) for batch_ind in range(100))

  def __len__(self):
    return 100

d = torch.utils.data.DataLoader(dataset = Dataset(), sampler = Sampler(None), num_workers = 0)
for i, x in enumerate(d):
    print(i)
    
```

**Comment by vadimkantorov** [comment]
@SsnL Hmm, maybe that's some other DataLoader bug.

Is this bit also fixed / or decided wontfix / or not relevant anymore?
"A particularly nasty deadlock may theoretically happen if all worker threads are blocked because data_queue exceeded its buffer size, and the main thread instead of removing the results from data_queue, fills up indices_queue and gets blocked because indices_queue is filled up. Given large enough batches (say with large lists in them transferred by value), this might happen during initial batch pre-fetching / priming."

**Comment by SsnL** [comment]
@vadimkantorov This is not relevant anymore. I believe that after #11985 , all (if not, then most) hang should be eliminated.

**Comment by vadimkantorov** [comment]
What would happen if:
0) Does the mp.Queue lock if we push a large object?
1) Priming loop puts a large enough index object so that the index_queues[i].put() locks, two times num_workers
2) Workers are all locked at data_queue.put()
3) Everything is locked, because number of batches under priming well exceeds the number of workers (that are all locked, since the main thread is locked at adding large indices objects)

This can only happen if the queue locks (and not throws) when a large object is put in.

**Comment by SsnL** [comment]
`mp.Queue`'s `put()` isn't blocking, so I don't think that it will happen.
